Get fullcontact profile update

rosnovsky opened the pull request:
### Description

FullContact had deprecated their v2 API at the end of December 2019. This PR updates FullContact rule with the new API URL and request format. It also unifies spelling of FullContact, and added a few comments.


### References

https://dashboard.fullcontact.com/api-ref#person-api-(v2)

> We suggest you upgrade to using our Person Enrich API as this endpoint will be discontinued by the end of December 2019. 

### Testing

No changes to any of the tests, they all still pass. Validation also passes just fine.

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [X] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used, if not `master`

shushen commented on src/rules/get-fullcontact-profile.js:
```diff
@@ -6,46 +6,50 @@
-  const FULLCONTACT_KEY = configuration.FULLCONTACT_KEY;
-  const SLACK_HOOK = configuration.SLACK_HOOK_URL;
+  // Get FullContact API key: https://dashboard.fullcontact.com/
+  const FULLCONTACT_KEY = 'YOUR FULLCONTACT API KEY'; 
```
Any particular reason why the support for `configuration` variables are removed here for `FULLCONTACT_KEY` and `SLACK_HOOK`?

rosnovsky replied:
Honestly, I just couldn't find a way to pass these as variables from elsewhere, and it looked more explicit and direct to just have these inline "instructions" for the user of what to put where in the rule to make it work. But this is subjective, of course.

shushen replied:
There is a global configuration object that is available for the rules and that's where the API keys and webhook URLs can be stored. Please see https://auth0.com/docs/rules/guides/configuration.

rosnovsky replied:
Oops, missed it completely. My bad. Just pushed an update.

shushen replied:
Awesome! Many thanks for the PR!

shushen commented on rules.json:
```diff
@@ -234,8 +234,8 @@
-        "description": "<p>This rule gets the user profile from FullContact using the e-mail (if available).</p>\n<p>If the information is immediately available (signaled by a <code>statusCode=200</code>), it adds a new property <code>fullContactInfo</code> to the user_metadata and returns. Any other conditions are ignored.</p>\n<p>See <a href=\"http://www.fullcontact.com/developer/docs/\">FullCo ...
-        "code": "function (user, context, callback) {\n  const FULLCONTACT_KEY = configuration.FULLCONTACT_KEY;\n  const SLACK_HOOK = configuration.SLACK_HOOK_URL;\n\n  const slack = require('slack-notify')(SLACK_HOOK);\n\n  // skip if no email\n  if (!user.email) return callback(null, user, context);\n\n  // skip if fullcontact metadata is already there\n  if (user.user_metadata && user.user_met ...
+        "description": "<p>This rule gets the user profile from FullContact using the e-mail (if available).</p>\n<p>If the information is immediately available (signaled by a <code>statusCode=200</code>), it adds a new property <code>fullcontact</code> to the user_metadata and returns. Any other conditions are ignored.</p>\n<p>See <a href=\"https://dashboard.fullcontact.com/api-ref\">FullContact ...
+        "code": "function (user, context, callback) {\n  // Get FullContact API key: https://dashboard.fullcontact.com/\n  const FULLCONTACT_KEY = 'YOUR FULLCONTACT API KEY'; \n\n  // Create a Slack app and get a hook URL: https://api.slack.com/messaging/webhooks#getting-started\n  const SLACK_HOOK = 'YOUR SLACK HOOK URL';\n\n  const slack = require('slack-notify')(SLACK_HOOK);\n\n  // skip if no ...
```
The generated JSON file needs be updated with the latest changes.

shushen commented on src/rules/get-fullcontact-profile.js:
```diff
@@ -6,46 +6,53 @@
-    qs: {
-      email:  user.email,
-      apiKey: FULLCONTACT_KEY
+  request.get('https://api.fullcontact.com/v3/person.enrich', {
```
@rosnovsky I had a second look at the FullContact V3 API and it appears to be a POST request: https://dashboard.fullcontact.com/api-ref#person-enrichment

```
curl -X POST https://api.fullcontact.com/v3/person.enrich \
-H "Authorization: Bearer {Your API Key}" \
-H "Content-Type: application/json" \
-d '{"email":"bart@fullcontact.com","twitter":"@bartlorang","fullName":"Bart Lorang"}'
```

joshcanhelp commented:
@rosnovsky - Is this something you're able to clean up? Or should we backlog a fix?

rosnovsky commented:
@joshcanhelp Damn them github notifications, sorry, missed this one :(

I'm working on this right now.

rosnovsky commented:
Fixed the request method (it's indeed POST, not GET), updated `rules.json`, updated the test file. I'm not sure how to properly check the expected header and URL in a test before returning `200` :/

joshcanhelp reviewed (changes requested):
A few changes and a question about keeping Slack in here. 

Also ... can you run `npm run format` to make sure the formatting is consistent?

Thank you for sticking with this!

joshcanhelp commented on src/rules/get-fullcontact-profile.js:
```diff
@@ -22,36 +28,34 @@ function getFullContactProfile(user, context, callback) {
+          error: error ? error.toString() : (response ? response.statusCode + ' ' + body : '')
+        }
+      });
+
```
A `console.log()` would help here. If an email is not found, you get a 404 and no profile.

joshcanhelp commented on src/rules/get-fullcontact-profile.js:
```diff
@@ -22,36 +28,34 @@ function getFullContactProfile(user, context, callback) {
+
+    // if we reach here, it means FullContact returned info and we'll add it to the metadata
+    user.user_metadata = user.user_metadata || {};
+    user.user_metadata.fullcontact = body;
```
Need to parse the JSON or else we just get a string. Something like:

```js
try {
  user.user_metadata.fullcontact = JSON.parse(body);
} catch (parseError) {
  console.log("Error parsing FullContact response: " + parseError.message);
  return callback(null, user, context);
}
```

joshcanhelp reviewed (changes requested):
Just tested. The call and everything works great, just one more processing step we need.

rosnovsky commented on src/rules/get-fullcontact-profile.js:
```diff
@@ -6,14 +6,20 @@
+  // Get FullContact API key: https://dashboard.fullcontact.com/
   const FULLCONTACT_KEY = configuration.FULLCONTACT_KEY;
+  
+  // Create a Slack app and get a hook URL: https://api.slack.com/messaging/webhooks#getting-started  
```
Makes total sense. I was wondering why is it even here, but thought it serves some purpose :)

rosnovsky commented:
Phew, so:
- Reverted `rules.json` to its previous state (I think)
- Updated the rule itself with `JSON.parse` try/catch
- Removed Slack from the rule (because why was it there in the first place?)
- Updated test file: the rule now receives a string instead of an object
- Ran `format` (hopefully, it worked)

joshcanhelp commented:
Thanks so much @rosnovsky! Made a few changes on my end:

- Added error handling to the API2 call
- We weren't actually saving the parsed object to the user profile
- I removed a property from the response that made the ID token very large. We still save the full response in Auth0

Thank you for all your work here! I'll get this merged in and released with the next round of Rule templates!

joshcanhelp commented:
One more ... the Rule now skips if the config key we need is not present.
